authors/apps/articles/views.py

```diff
@@ -50,9 +51,60 @@ class ArticleDetailsView(RetrieveUpdateAPIView):
     serializer_class = GetArticleSerializer
     lookup_field = 'slug'
 
+    def get(self, request, slug, *args, **kwargs):
+        """Retrieve a single article"""
+        article = get_object_or_404(Article, slug=self.kwargs["slug"])
+        serializer = self.serializer_class(article)
+        return set_favorite_status(serializer, self.request.user.id)
+
 
 class DeleteArticle(UpdateAPIView):
     permission_classes = (IsAuthorOrReadOnly,)
     queryset = Article.objects.filter(is_deleted=False)
     serializer_class = DeleteArticleSerializer
     lookup_field = 'slug'
+
+
+class FaveArticle(ListCreateAPIView):
+    """Favorite an Article."""
+    permission_classes = (IsAuthenticated,)
+    serializer_class = ArticleSerializer
+
+    def put(self, request, slug):
+        try:
+            # get an article that matches the slug specified
+            # and return a message if the article does not exist
+            article = Article.objects.get(slug=slug)
+        except ObjectDoesNotExist:
+            raise NotFound("This article does not exist")
+
+        # check if the article has been favorited by the current user
+        # if it is favorited we unfavorite it
+        if article in Article.objects.filter(
+                                        favorite=request.user):
+            article.favorite.remove(request.user)
+
+        # If it is not favorited we favorite the article
+        else:
+            article.favorite.add(request.user)
+
+        serializer = self.serializer_class(article,
+                                           context={'request': request},
+                                           partial=True)
+        return set_favorite_status(serializer, self.request.user.id)
+
+
+def set_favorite_status(data, user):
+        """Set the favorite value to either True or False"""
+        # Check if the user has favorited the article
+        value = [value for value in data.data['favorite'] if
+                 user in data.data['favorite']]
+        # Article has not been favorited, set favorite to False
```

Comment:
Article has been favorited you mean?

Reply:
I have refactored the code and changes this